Notebook tool implementation

Kiwi-520 opened the pull request:
## TLDR

This pull request adds a built-in NotebookEditTool to Gemini CLI, resolving #6930.  
It enables safe, reliable editing of Jupyter notebooks (.ipynb) with dedicated operations for adding, editing, deleting, moving cells, and clearing outputs.  
This approach prevents JSON corruption, ensures data safety, and integrates directly with the CLI—no external dependencies required.

## Dive Deeper

- Implements NotebookEditTool as a BaseDeclarativeTool with async file operations and crypto-secure cell IDs.
- Supports 5 core notebook operations: add_cell, edit_cell, delete_cell, move_cell, clear_outputs.
- Validates notebook structure before and after every operation to prevent corruption.
- Comprehensive error handling: prevents deleting the last cell, checks for invalid indices/IDs, and ensures only valid notebook files are edited.
- Fully type-safe (TypeScript) and tested with 12 unit tests covering all features and edge cases.
- No external servers or background processes—works out of the box in any IDE.

## Reviewer Test Plan

1. Pull the branch and install dependencies:
   npm install
   npm run build

2. Run the test suite:
   cd packages/core
   npx vitest run src/tools/notebook-edit.test.ts

3. Manual testing:
   - Use a sample .ipynb file (e.g., helloworld_maths.ipynb).
   - Use Gemini CLI with prompts like:
     - Add a markdown cell at the end
     - Edit the first code cell
     - Delete a cell by index or ID
     - Move a cell to a new position
     - Clear all outputs
   - Confirm the notebook remains valid and changes are as expected.

4. Edge cases:
   - Try deleting all cells (should prevent deleting the last cell).
   - Attempt operations with invalid indices or IDs (should return clear errors).
   - Edit a malformed notebook file (should not corrupt the file).

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ✅  | ✅  |
| npx      | ✅  | ✅  | ✅  |
| Docker   | ✅  | ✅  | ✅  |
| Podman   | ✅  | -   | -   |
| Seatbelt | ✅  | -   | -   |

## Linked issues / bugs

Resolves #6930

Kiwi-520 commented:
Thank you for the detailed review and suggestions!
I’ve made the recommended changes to the move_cell logic and the handling of empty cell content for both add and edit operations.
Please let me know if there’s anything else I should address.
Appreciate your help and feedback!

Kiwi-520 commented:
Hi @google-gemini/gemini-cli-maintainers 👋  

I’ve addressed the review feedback, resolved the `move_cell` logic issue, and fixed empty cell handling as suggested.  
The CLA check is now ✅, and the branch is up to date with `main`.  

Could you please review and approve the workflows when you have a chance?  
Thank you for your time and support! 🙏

ivan-stout commented:
Hi @jacob314,

Thank you for the thoughtful feedback. Regarding your point:

> However this should probably be an extension to Gemini CLI or perhaps an MCP server rather than directly integrated into Gemini CLI.

This sparked a broader idea that I believe could be a significant feature for the project, and I was hoping you could point me to the right forum (perhaps a GitHub Discussion or a new issue) to propose it.

To give some context, I'm in the process of rolling out `gemini-cli` to a team of a dozen developers. A native session export feature would be incredibly powerful for our collaboration, and we believe it could be a key differentiator from competing projects.

The feature we envision is a native command:
`gemini session export my_session.ipynb`

The primary benefits for a team like ours would be:

*   **High-Fidelity Collaboration:** It would allow us to share sessions that capture not just text and code, but also the structured outputs, turning a conversation into a rich, reviewable document.
*   **Reproducible Artifacts:** This would turn a transient CLI session into a persistent artifact that can be archived, shared, and analyzed, which is invaluable for tutorials, bug reports, and knowledge sharing.
*   **Ecosystem Integration:** It would instantly make `gemini-cli` sessions compatible with the vast Jupyter ecosystem, allowing them to be opened and enriched in tools our team already uses, like VS Code, JupyterLab, and Google Colab.

We feel that the `.ipynb` format is uniquely suited for this, as it's the de facto open standard for combining narrative, code, and outputs.

Again, I'm happy to take this discussion to a more appropriate channel. Please let me know what the best next step is.

Thank you for your time and consideration.

---
*sent from gemini-cli*

Kiwi-520 commented:
I have Merge Conflicts Resolved & Build Fixed

All merge conflicts have been resolved across the repository.

Core, a2a-server, test-utils, and VSCode companion packages build successfully.

prompts.ts compilation issue fully fixed.

TypeScript configuration optimized (ES2022+, esModuleInterop, skipLibCheck).

Remaining issues: 11 minor syntax errors in github.ts and useExtensionUpdates.ts (CLI package only). These do not break the build but will need follow-up.

Ready for Review / Approval

mattKorwel commented:
Hi, I see quite a few changed files that aren't a part of this work; all the .github files to be specific. I even see a release run? what's the status of this here: https://github.com/google-gemini/gemini-cli/actions/workflows/release-manual.yml

Kiwi-520 commented:
Hi @mattKorwel,

I’d like to apologize for the accidental changes I introduced to the .github files earlier. As a beginner to open source, I didn’t realize these workflows should not be modified.

I’ve now reverted all .github changes back to match main, so there should be no release workflow triggers or unrelated file edits in this PR anymore.

Thank you for pointing this out and for your patience I’m learning a lot through this contribution. Please let me know if there’s anything else I should fix or adjust. 🙏

ivan-stout commented:
> `gemini session export my_session.ipynb`

I was just reading through the discussion on PR #5342 and saw that @westurner had already suggested using `.ipynb` for exports as a way to capture prompts, outputs, and metadata (see comment: https://github.com/google-gemini/gemini-cli/pull/5342#issuecomment-3240424195).
 
It seems there's a lot of overlap between that idea and the proposal in this PR. Both conversations highlight a user need for a richer, more structured way to handle notebooks, whether that's creating them via session export or editing them directly.

Just wanted to connect these two threads, as PR #5342 has been open for a bit. It's also worth noting that the original `/export` command in #5342 was implemented as a native feature.

Kiwi-520 commented:
Thank you all for the guidance, it’s been very helpful. I’ll go ahead and close this PR for now and start working on re-implementing the NotebookEditTool as a Gemini CLI extension or MCP server, as suggested. Once the extension is ready, I’ll open a new, focused PR for review. Thanks again for your all your support and patience.
